File: Chapter11/C#/SystemBrowser/Workers/Content/ContentSearchResults.cs
using System;
using System.IO;
using System.Collections;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Windows.Forms;

namespace SystemBrowser
{
	/// <summary>
  /// ContentSearchResults is used to show the results of a file search
  /// </summary>
	public class ContentSearchResults : System.Windows.Forms.UserControl
	{
    #region Controls
    private System.Windows.Forms.ListView listViewFiles;
    private System.Windows.Forms.ColumnHeader columnHeaderName;
    private System.Windows.Forms.ColumnHeader columnHeaderSize;
    private System.Windows.Forms.ColumnHeader columnHeaderType;
    private System.Windows.Forms.ColumnHeader columnHeaderDateModified;
    private System.Windows.Forms.ColumnHeader columnHeaderInFolder;
    private System.ComponentModel.IContainer components;
    private System.Windows.Forms.ImageList imageListFiles;
    #endregion

    // ImageList indexes, for images placed next to files found
    const int ImageFile = 0;

    public ContentSearchResults()
		{
			// This call is required by the Windows.Forms Form Designer.
			InitializeComponent();
    }

    // add a file to the list of files found
    public void Add(string theFilePath, int theCurrentCount)
    {
      if (!File.Exists(theFilePath)) return;

      string name = Path.GetFileName(theFilePath);
      ListViewItem item = new ListViewItem(name, ImageFile);
      item.Tag = theFilePath;
      FileInfo fileInfo = new FileInfo(theFilePath);
      string folder = fileInfo.DirectoryName;
      item.SubItems.Add(folder);
      long length = fileInfo.Length / 1024;
      if (length < 1)
        length = 1;
      string size = string.Format("{0:#,##0} KB", length);
      string type = fileInfo.Extension;
      if (type.StartsWith("."))
        type = type.Substring(1);
      DateTime dateModified = fileInfo.LastWriteTime;
      item.SubItems.Add(size);
      item.SubItems.Add(type);
      item.SubItems.Add(dateModified.ToString("yyyy-MM-dd HH:mm:ss"));
      listViewFiles.Items.Add(item);
    }

    // organize file list showing only names
    public void ShowIcons()
    {
      listViewFiles.View = View.SmallIcon;
    }

    // organize file list showing details
    public void ShowDetails()
    {
      listViewFiles.View = View.Details;
    }

    public void Clear()
    {
      listViewFiles.Items.Clear();
    }

    /// <summary> 
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if(components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Component Designer generated code
		/// <summary> 
		/// Required method for Designer support - do not modify 
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
      this.components = new System.ComponentModel.Container();
      System.Resources.ResourceManager resources = new System.Resources.ResourceManager(typeof(ContentSearchResults));
      this.listViewFiles = new System.Windows.Forms.ListView();
      this.columnHeaderName = new System.Windows.Forms.ColumnHeader();
      this.columnHeaderInFolder = new System.Windows.Forms.ColumnHeader();
      this.columnHeaderSize = new System.Windows.Forms.ColumnHeader();
      this.columnHeaderType = new System.Windows.Forms.ColumnHeader();
      this.columnHeaderDateModified = new System.Windows.Forms.ColumnHeader();
      this.imageListFiles = new System.Windows.Forms.ImageList(this.components);
      this.SuspendLayout();
      // 
      // listViewFiles
      // 
      this.listViewFiles.BorderStyle = System.Windows.Forms.BorderStyle.FixedSingle;
      this.listViewFiles.Columns.AddRange(new System.Windows.Forms.ColumnHeader[] {
                                                                                    this.columnHeaderName,
                                                                                    this.columnHeaderInFolder,
                                                                                    this.columnHeaderSize,
                                                                                    this.columnHeaderType,
                                                                                    this.columnHeaderDateModified});
      this.listViewFiles.Dock = System.Windows.Forms.DockStyle.Fill;
      this.listViewFiles.HideSelection = false;
      this.listViewFiles.Location = new System.Drawing.Point(0, 0);
      this.listViewFiles.Name = "listViewFiles";
      this.listViewFiles.Size = new System.Drawing.Size(554, 364);
      this.listViewFiles.SmallImageList = this.imageListFiles;
      this.listViewFiles.TabIndex = 1;
      this.listViewFiles.View = System.Windows.Forms.View.Details;
      this.listViewFiles.Resize += new System.EventHandler(this.listViewFiles_Resize);
      this.listViewFiles.SelectedIndexChanged += new System.EventHandler(this.listViewFiles_SelectedIndexChanged);
      // 
      // columnHeaderName
      // 
      this.columnHeaderName.Text = "Name";
      this.columnHeaderName.Width = 143;
      // 
      // columnHeaderInFolder
      // 
      this.columnHeaderInFolder.Text = "In Folder";
      this.columnHeaderInFolder.Width = 162;
      // 
      // columnHeaderSize
      // 
      this.columnHeaderSize.Text = "Size";
      this.columnHeaderSize.TextAlign = System.Windows.Forms.HorizontalAlignment.Right;
      this.columnHeaderSize.Width = 76;
      // 
      // columnHeaderType
      // 
      this.columnHeaderType.Text = "Type";
      // 
      // columnHeaderDateModified
      // 
      this.columnHeaderDateModified.Text = "Date Modified";
      this.columnHeaderDateModified.Width = 109;
      // 
      // imageListFiles
      // 
      this.imageListFiles.ColorDepth = System.Windows.Forms.ColorDepth.Depth24Bit;
      this.imageListFiles.ImageSize = new System.Drawing.Size(16, 16);
      this.imageListFiles.ImageStream = ((System.Windows.Forms.ImageListStreamer)(resources.GetObject("imageListFiles.ImageStream")));
      this.imageListFiles.TransparentColor = System.Drawing.Color.Transparent;
      // 
      // ContentSearchResults
      // 
      this.Controls.Add(this.listViewFiles);
      this.Name = "ContentSearchResults";
      this.Size = new System.Drawing.Size(554, 364);
      this.ResumeLayout(false);

    }
		#endregion

    #region Event Handlers
    private void listViewFiles_SelectedIndexChanged(object sender, System.EventArgs e)
    {
      if (listViewFiles.SelectedItems.Count == 0)
      {
        string s = string.Format("{0} objects", listViewFiles.Items.Count);
        FireMessage(s);  // no file selected
      }
      else
      {
        ListViewItem item = listViewFiles.SelectedItems[0];
        string folder = item.SubItems[1].Text;
        FireMessage("In folder: " + folder);
      }
    }

    private void listViewFiles_Resize(object sender, System.EventArgs e)
    {
      columnHeaderDateModified.Width = -2;  // autosize last column
    }
    #endregion

    #region Events
    public delegate void MessageHandler(string theMessage);
    public event MessageHandler OnMessage;
    private void FireMessage(string theMessage)
    {
      if (OnMessage != null)
        OnMessage(theMessage);
    }
    #endregion
  }
}

File: Chapter11/VB/SystemBrowser/Workers/Content/ContentSearchResults.vb
Imports System.IO

Public Class ContentSearchResults
  Inherits System.Windows.Forms.UserControl

  ' ImageList indexes, for images placed next to files found
  Const ImageFile As Integer = 0

  ' add a file to the list of files found
  Public Sub Add(ByVal theFilePath As String, ByVal theCurrentCount As Integer)
    If Not File.Exists(theFilePath) Then Return

    Dim name As String = Path.GetFileName(theFilePath)
    Dim item As New ListViewItem(name, ImageFile)
    item.Tag = theFilePath
    Dim fi As New FileInfo(theFilePath)
    Dim folder As String = fi.DirectoryName
    item.SubItems.Add(folder)
    Dim length As Long = CLng(fi.Length / 1024)
    If (length < 1) Then
      length = 1
    End If
    Dim size As String = String.Format("{0:#,##0} KB", length)
    Dim type As String = fi.Extension
    If type.StartsWith(".") Then
      type = type.Substring(1)
    End If
    Dim dateModified As DateTime = fi.LastWriteTime
    item.SubItems.Add(size)
    item.SubItems.Add(type)
    item.SubItems.Add(dateModified.ToString("yyyy-MM-dd HH:mm:ss"))
    listViewFiles.Items.Add(item)
  End Sub

  ' organize file list showing only names
  Public Sub ShowIcons()
    listViewFiles.View = View.SmallIcon
  End Sub

  ' organize file list showing details
  Public Sub ShowDetails()
    listViewFiles.View = View.Details
  End Sub

  Public Sub Clear()
    listViewFiles.Items.Clear()
  End Sub

#Region " Windows Form Designer generated code "

  Public Sub New()
    MyBase.New()

    'This call is required by the Windows Form Designer.
    InitializeComponent()

    'Add any initialization after the InitializeComponent() call

  End Sub

  'UserControl overrides dispose to clean up the component list.
  Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
    If disposing Then
      If Not (components Is Nothing) Then
        components.Dispose()
      End If
    End If
    MyBase.Dispose(disposing)
  End Sub

  'Required by the Windows Form Designer
  Private components As System.ComponentModel.IContainer

  'NOTE: The following procedure is required by the Windows Form Designer
  'It can be modified using the Windows Form Designer.  
  'Do not modify it using the code editor.
  Friend WithEvents imageListFiles As System.Windows.Forms.ImageList
  Friend WithEvents listViewFiles As System.Windows.Forms.ListView
  Friend WithEvents columnHeaderName As System.Windows.Forms.ColumnHeader
  Friend WithEvents columnHeaderInFolder As System.Windows.Forms.ColumnHeader
  Friend WithEvents columnHeaderSize As System.Windows.Forms.ColumnHeader
  Friend WithEvents columnHeaderType As System.Windows.Forms.ColumnHeader
  Friend WithEvents columnHeaderDateModified As System.Windows.Forms.ColumnHeader
  <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
    Me.components = New System.ComponentModel.Container
    Dim resources As System.Resources.ResourceManager = New System.Resources.ResourceManager(GetType(ContentSearchResults))
    Me.imageListFiles = New System.Windows.Forms.ImageList(Me.components)
    Me.listViewFiles = New System.Windows.Forms.ListView
    Me.columnHeaderName = New System.Windows.Forms.ColumnHeader
    Me.columnHeaderInFolder = New System.Windows.Forms.ColumnHeader
    Me.columnHeaderSize = New System.Windows.Forms.ColumnHeader
    Me.columnHeaderType = New System.Windows.Forms.ColumnHeader
    Me.columnHeaderDateModified = New System.Windows.Forms.ColumnHeader
    Me.SuspendLayout()
    '
    'imageListFiles
    '
    Me.imageListFiles.ColorDepth = System.Windows.Forms.ColorDepth.Depth24Bit
    Me.imageListFiles.ImageSize = New System.Drawing.Size(16, 16)
    Me.imageListFiles.ImageStream = CType(resources.GetObject("imageListFiles.ImageStream"), System.Windows.Forms.ImageListStreamer)
    Me.imageListFiles.TransparentColor = System.Drawing.Color.Transparent
    '
    'listViewFiles
    '
    Me.listViewFiles.BorderStyle = System.Windows.Forms.BorderStyle.FixedSingle
    Me.listViewFiles.Columns.AddRange(New System.Windows.Forms.ColumnHeader() {Me.columnHeaderName, Me.columnHeaderInFolder, Me.columnHeaderSize, Me.columnHeaderType, Me.columnHeaderDateModified})
    Me.listViewFiles.Dock = System.Windows.Forms.DockStyle.Fill
    Me.listViewFiles.HideSelection = False
    Me.listViewFiles.Location = New System.Drawing.Point(0, 0)
    Me.listViewFiles.Name = "listViewFiles"
    Me.listViewFiles.Size = New System.Drawing.Size(554, 364)
    Me.listViewFiles.SmallImageList = Me.imageListFiles
    Me.listViewFiles.TabIndex = 2
    Me.listViewFiles.View = System.Windows.Forms.View.Details
    '
    'columnHeaderName
    '
    Me.columnHeaderName.Text = "Name"
    Me.columnHeaderName.Width = 143
    '
    'columnHeaderInFolder
    '
    Me.columnHeaderInFolder.Text = "In Folder"
    Me.columnHeaderInFolder.Width = 162
    '
    'columnHeaderSize
    '
    Me.columnHeaderSize.Text = "Size"
    Me.columnHeaderSize.TextAlign = System.Windows.Forms.HorizontalAlignment.Right
    Me.columnHeaderSize.Width = 76
    '
    'columnHeaderType
    '
    Me.columnHeaderType.Text = "Type"
    '
    'columnHeaderDateModified
    '
    Me.columnHeaderDateModified.Text = "Date Modified"
    Me.columnHeaderDateModified.Width = 109
    '
    'ContentSearchResults
    '
    Me.Controls.Add(Me.listViewFiles)
    Me.Name = "ContentSearchResults"
    Me.Size = New System.Drawing.Size(554, 364)
    Me.ResumeLayout(False)

  End Sub

#End Region

#Region "Event Handlers"
  Private Sub listViewFiles_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles listViewFiles.SelectedIndexChanged
    If listViewFiles.SelectedItems.Count = 0 Then
      Dim s As String = String.Format("{0} objects", listViewFiles.Items.Count)
      FireMessage(s)  ' no file selected
    Else
      Dim item As ListViewItem = listViewFiles.SelectedItems(0)
      Dim folder As String = item.SubItems(1).Text
      FireMessage("In folder: " + folder)
    End If
  End Sub

  Private Sub listViewFiles_Resize(ByVal sender As Object, ByVal e As System.EventArgs) Handles listViewFiles.Resize
    columnHeaderDateModified.Width = -2  ' autosize last column
  End Sub

#End Region

#Region "Events"
  Public Event OnMessage(ByVal theMessage As String)
  Private Sub FireMessage(ByVal theMessage As String)
    RaiseEvent OnMessage(theMessage)
  End Sub
#End Region

End Class
